File: 0001_simple_db_in_memory_index/python/simple_db_in_memory_index.py
import asyncio
import functools
import json


class _Index:
    def __init__(self):
        self._idx_map = {}
        self._cursor = 0

    async def add_next(self, line: str, key=None):
        if key is None:
            key, _  = line.split(',', 1)
        length = len(line)
        self._idx_map[key] = (self._cursor, length)
        self._cursor += length

    async def get(self, key):
        return self._idx_map.get(key)


def _load_index(func):
    @functools.wraps(func)
    async def wrapper(self, *args, **kwargs):
        tasks = []
        if not self._index_loaded:
            try:
                with open(self.filename, 'r') as f:
                    for line in f:
                        tasks.append(asyncio.create_task(self._index.add_next(line)))
            except FileNotFoundError:
                pass
            self._index_loaded = True
            for task in tasks:
                await task

        return await func(self, *args, **kwargs)

    return wrapper


class SimpleDbInMemoryIndex:
    def __init__(self, filename='database'):
        self.filename = filename
        self._index = _Index()
        self._index_loaded = False

    @_load_index
    async def set(self, key, json_dict):
        line = f'{key},{json.dumps(json_dict)}\n'
        task = asyncio.create_task(self._index.add_next(line, key=key))
        with open(self.filename, 'a') as f:
            f.write(line)
        await task

    @_load_index
    async def get(self, key):
        try:
            offset, length = await self._index.get(key) 
        except TypeError:
            return None

        with open(self.filename, 'r') as f:
            f.seek(offset)
            return json.loads(f.read(length).split(',', 1)[1].strip())

File: 0001_simple_db_in_memory_index/cpp/simple_db_in_memory_index.cpp
#include "simple_db_in_memory_index.h"
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <utility>

_Index::_Index() : _cursor(0) {}

void _Index::add_next(const std::string &line, const std::string &key) {
  std::string actual_key = key;
  if (key.empty()) {
    std::istringstream iss(line);
    getline(iss, actual_key, ',');
  }
  size_t length = line.length();
  _idx_map[actual_key] = {_cursor, length};
  _cursor += length;
}

std::pair<long, size_t> _Index::get(const std::string &key) const {
  auto it = _idx_map.find(key);
  if (it != _idx_map.end()) {
    return it->second;
  }
  throw std::runtime_error("Key not found");
}

SimpleDbInMemoryIndex::SimpleDbInMemoryIndex(const std::string &filename)
    : filename(filename), _index() {
  std::ifstream file(filename);
  if (!file.is_open()) {
    return;
  }

  std::string line;
  while (getline(file, line)) {
    _index.add_next(line + "\n");
  }
}

void SimpleDbInMemoryIndex::set(const std::string &key,
                                const nlohmann::json &json_dict) {
  std::string line = key + "," + json_dict.dump() + "\n";
  std::ofstream file(filename, std::ios::app);
  file << line;
  file.close();
  _index.add_next(line, key);
}

nlohmann::json SimpleDbInMemoryIndex::get(const std::string &key) {
  std::pair<long, size_t> offset_length;
  try {
    offset_length = _index.get(key);
  } catch (const std::runtime_error &) {
    return nullptr;
  }

  std::ifstream file(filename);
  file.seekg(offset_length.first);
  std::string line;
  line.resize(offset_length.second);
  file.read(&line[0], offset_length.second);
  file.close();
  return nlohmann::json::parse(line.substr(line.find(',') + 1));
}